### PR TITLE
Join reordering stats11

### PR DESCRIPTION
**Pull request by rschlussel-zz**
supersedes #584

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+            // If we're doing a cross join, use all output symbols from the inputs and add a project node
+            // on top
+            List<Symbol> joinOutputSymbols = sortedOutputSymbols;
+            if (joinConditions.isEmpty() && joinFilters.isEmpty()) {
+                joinOutputSymbols = Stream.concat(left.getOutputSymbols().stream(), right.getOutputSymbols().stream())
+                        .collect(toImmutableList());
+            }
+
+            JoinEnumerationResult result = setJoinNodeProperties(new JoinNode(
+                    idAllocator.getNextId(),
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:302]: id counter (int) might not be enough in long term

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+                memo.put(multiJoinKey, bestResult);
+            }
+            if (bestResult.planNode.isPresent()) {
+                log.debug("Least cost join was: " + bestResult.planNode.get().toString());
+            }
+            return bestResult;
+        }
+
+        /**
+         * This method generates all the ways of dividing  totalNodes into two sets
```
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java]: double space after 'dividing'

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+        {
+            Set<Symbol> leftSymbols = leftSources.stream()
+                    .flatMap(node -> node.getOutputSymbols().stream())
+                    .collect(toImmutableSet());
+            Set<Symbol> rightSymbols = rightSources.stream()
+                    .flatMap(node -> node.getOutputSymbols().stream())
+                    .collect(toImmutableSet());
+            ImmutableList.Builder<Expression> joinPredicatesBuilder = ImmutableList.builder();
+
+            // add join conjucts that were not used for inference
```
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java]: s/conjucts/conjuncts

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+            stream(EqualityInference.nonInferrableConjuncts(allFilter).spliterator(), false)
+                    .map(conjuct -> allInference.rewriteExpression(conjuct, symbol -> leftSymbols.contains(symbol) || rightSymbols.contains(symbol)))
```
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java]: s/conjuct/conjunct

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+        }
+
+        private JoinEnumerationResult chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols)
+        {
+            Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
+            JoinEnumerationResult bestResult = memo.get(multiJoinKey);
+            if (bestResult == null) {
+                checkState(sources.size() > 1, "sources size is less than or equal to one");
+                ImmutableList.Builder<JoinEnumerationResult> resultBuilder = ImmutableList.builder();
+                Set<Set<Integer>> partitions = generatePartitions(sources.size()).collect(toImmutableSet());
```
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:162]: IIUC, we could possibly spare lots of object allocations if iterating over partitions based on BitSets instead of `Set<Integer>`. However, we would need something like https://gist.github.com/avianey/3ea9dbc64489e4058d422d3b270644a0 .

File: presto-main/src/main/java/com/facebook/presto/testing/TestingLookup.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
+public class TestingLookup
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingLookup.java:35]: see my approach: https://github.com/Teradata/presto/pull/605/commits/bba57cac0e4a3f76931a38b359f47a9193253c20
and https://github.com/Teradata/presto/pull/605/commits/0eb6a5f541a178e6b78027323d447dc46d7e082a#diff-a5cd1e00ca43c92920736a87b29dbcfb
What do you think? I think it's more cleaner with composite approach

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java (hunk 5 above)
Comment (rschlussel-zz) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:162]: Let's pick this up for future work.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java (hunk 1 above)
Comment (rschlussel-zz) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:302]: we can change it to a long once we run into that issue.

File: presto-main/src/main/java/com/facebook/presto/testing/TestingLookup.java (hunk 6 above)
Comment (rschlussel-zz) [reply presto-main/src/main/java/com/facebook/presto/testing/TestingLookup.java:35]: agreed

**Comment by rschlussel-zz** [comment]
merged to the sprint branch. @sopel39 your pr can go on top.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+            // sort output symbols so that the left input symbols are first
+            List<Symbol> sortedOutputSymbols = Stream.concat(left.getOutputSymbols().stream(), right.getOutputSymbols().stream())
+                    .filter(outputSymbols::contains)
+                    .collect(toImmutableList());
+
+            // Cross joins can't filter symbols as part of the join
+            // If we're doing a cross join, use all output symbols from the inputs and add a project node
+            // on top
+            List<Symbol> joinOutputSymbols = sortedOutputSymbols;
+            if (joinConditions.isEmpty() && joinFilters.isEmpty()) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:296]: this is no longer needed since we don't allow cross joins

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+                    left,
+                    right,
+                    joinConditions,
+                    joinOutputSymbols,
+                    joinFilters.isEmpty() ? Optional.empty() : Optional.of(and(joinFilters)),
+                    Optional.empty(),
+                    Optional.empty(),
+                    Optional.empty()));
+
+            if (!joinOutputSymbols.equals(sortedOutputSymbols)) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:313]: I don't think it's needed either

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java [hunk 9]
```diff
@@ -111,6 +111,9 @@ public PlanNode visitDelete(DeleteNode node, RewriteContext<Void> context)
 
         private JoinNode.DistributionType getTargetJoinDistributionType(JoinNode node)
         {
+            if (node.getDistributionType().isPresent()) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java:114]: this is unrelated change to this commit. Make a (small) commit out of it.

**Review by sopel39 — COMMENTED** [review]
Algorithm LGTM. I will make final pass when we fix tests
